Fix payments with custom order status made via payment or setup intents

reykjalin opened the pull request:
# Changes proposed in this Pull Request:

Supercedes #1499 

This filter is used to control which order statuses allow a payment to be processed in the context of 3DS payments and the associated webhooks. By default the only statuses treated as such are `pending` and `failed`.

This is mostly useful when merchants assign a custom status to an order before payment processing begins. Before this commit a custom order status would capture the payment, but the order status wasn't updated.

# Testing instructions

> **Note:** Webhooks must be configured correctly for the testing to work correctly.

> **Note:** You must install the [WooCommerce Blocks](https://wordpress.org/plugins/woo-gutenberg-products-block/) plugin for the **Draft** order status to be available.

1. Make sure payments are captured immediately by going to **WooCommerce > Settings > Payments > Stripe – Credit Card (Stripe)**, make sure **Capture** is toggled on, and click **Save changes**.
2. Install and activate the [Code Snippets](https://wordpress.org/plugins/code-snippets/) plugin.
3. Go to **Snippets > Add new**
4. Copy the following snippet into the **Code** text area. You can choose any name you want.

```php
function add_draft_to_allowed_payment_processing_statuses( $statuses ) {
	if ( ! in_array( 'checkout-draft', $statuses ) ) {
		$statuses[] = 'checkout-draft';
	}
	return $statuses;
}
add_filter( 'wc_stripe_allowed_payment_processing_statuses', 'add_draft_to_allowed_payment_processing_statuses' );
```

<img width="967" alt="image" src="https://user-images.githubusercontent.com/13835680/128262187-2b0bc3fd-d263-4051-b3f5-86686cbdb2a6.png">

5. Make sure **Run snippet everywhere** is toggled
6. Click **Save changes and activate**
7. Go to **WooCommerce > Orders** and click **Add Order**
8. Set the order status to **Draft**
9. Select the customer for your current account — or any account you have access to on your site. The customer details should be filled in automatically.

<img width="893" alt="image" src="https://user-images.githubusercontent.com/13835680/128262549-3eafe408-c729-4b69-84bf-b52a9a792e0b.png">

10. Click **Add item(s)** followed by **Add product(s)** and add any product to the order.

<img width="893" alt="image" src="https://user-images.githubusercontent.com/13835680/128262620-ec5ed835-071a-464f-8cb6-2bd010a0cc18.png">

<img width="893" alt="image" src="https://user-images.githubusercontent.com/13835680/128262716-d0c82e8a-8e87-462c-a1d3-a63534dd9a99.png">

11. Click **Create**
12. Click **Customer payment page**
13. Pay for the order with a 3DS card: `4000000000003220`
14. Go back to the order details page
15. Make sure the order status has changed to **Processing** (or **completed** for digital orders).
16. Turn off immediate payment capture by going to **WooCommerce > Settings > Payments > Stripe – Credit Card (Stripe)**, make sure **Capture** is toggled off, and click **Save changes**. 
17. Repeat steps 2-14 (create a draft order, pay through the **Customer payment page** page)
18. Make sure the order status has changed to **On hold**.
19. Make sure there is a note that starts with `Stripe charge authorized (Charge ID: ch_xyz).`
20. Change the payment status to **Processing**
21. Make sure a note has appeared that says `Stripe charge complete (Charge ID: ch_3JKqQ7KFePZ3MR450VtiEw3i)`

Repeat these testing instructions with the following conditions:

1. Use a non-3DS card and make sure everything behaves the same.
2. Add a product to the draft order that has stock and make sure the stock management behaves in the following ways:
    - Stock is not reduced when the order is created (after step 11)
    - (For immediate capture) stock is reduced after payment (after step 13)
    - (For manual capture) stock is reduced after authorization (after step 20)
    - **Tip:** To add stock to a product go to **Products**, select any product, go to **Inventory**, toggle **Manage stock?** on, change **Stock quantity** to a positive number, and click **Update**.


-------------------
- [x] Make sure your changes respect [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- [x] Did you make changes, or create a **new .js file**? If **Gruntfile.js** exists in the repo, make sure to run `grunt`.

Commit:
Add allowed_payment_processing_statuses filter

This filter is used to control which order statuses allow a payment to
be processed in the context of 3DS payments and the associated webhooks.
By default the only statuses treated as such are 'pending' and 'failed'.

This is mostly useful when merchants assign a custom status to an order
before payment processing begins. Before this commit a custom order
status would capture the payment, but the order status wasn't updated.

See PR #1499 for more information.

Example usage:

    public function allow_reserved_order_status( $statuses ) {
        // This assumes a 'reserved' order status slug already exists.
        if ( ! in_array( 'reserved', $statuses ) ) {
            $statuses[] = 'reserved';
        }
        return $statuses;
    }
    add_filter(
        'wc_stripe_allowed_payment_processing_statuses',
        'allow_reserved_order_status'
    );

Commit:
add changelog entry

asumaran reviewed (commented):
@reykjalin it looks like we need to apply the filter [here](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/develop/includes/class-wc-stripe-webhook-handler.php#L300) as well since we are processing payment possibly in other allowed statuses other than those two ("pending", "failed"). 

Actually, it seems there are more places where we might need to add the filter:


```
woocommerce-gateway-stripe/includes/class-wc-gateway-stripe.php:
 1116:        [ 'pending', 'failed' ]

woocommerce-gateway-stripe/includes/class-wc-stripe-webhook-handler.php:
  300:      $statuses = [ 'pending', 'failed' ];
  752:        [ 'pending', 'failed' ]
  804:        [ 'pending', 'failed' ]

woocommerce-gateway-stripe/includes/abstracts/abstract-wc-stripe-payment-gateway.php:
  474:      if ( $order->has_status( [ 'pending', 'failed' ] ) ) {

woocommerce-gateway-stripe/includes/payment-methods/class-wc-gateway-stripe-alipay.php:
  284:      $statuses = [ 'pending', 'failed' ];

woocommerce-gateway-stripe/includes/payment-methods/class-wc-gateway-stripe-bancontact.php:
  273:      if ( $order->has_status( [ 'pending', 'failed' ] ) ) {

woocommerce-gateway-stripe/includes/payment-methods/class-wc-gateway-stripe-eps.php:
  272:      if ( $order->has_status( [ 'pending', 'failed' ] ) ) {

woocommerce-gateway-stripe/includes/payment-methods/class-wc-gateway-stripe-giropay.php:
  272:      if ( $order->has_status( [ 'pending', 'failed' ] ) ) {

woocommerce-gateway-stripe/includes/payment-methods/class-wc-gateway-stripe-ideal.php:
  272:      if ( $order->has_status( [ 'pending', 'failed' ] ) ) {

woocommerce-gateway-stripe/includes/payment-methods/class-wc-gateway-stripe-multibanco.php:
  379:      if ( $order->has_status( [ 'pending', 'failed' ] ) ) {

woocommerce-gateway-stripe/includes/payment-methods/class-wc-gateway-stripe-p24.php:
  269:      if ( $order->has_status( [ 'pending', 'failed' ] ) ) {

woocommerce-gateway-stripe/includes/payment-methods/class-wc-gateway-stripe-sepa.php:
  408:      if ( $order->has_status( [ 'pending', 'failed' ] ) ) {

woocommerce-gateway-stripe/includes/payment-methods/class-wc-gateway-stripe-sofort.php:
  285:      if ( $order->has_status( [ 'pending', 'failed' ] ) ) {
```

reykjalin commented:
> @reykjalin it looks like we need to apply the filter here as well since we are processing payment possibly in other allowed statuses other than those two ("pending", "failed").

I don't think that's necessary. The order status is updated before we enter the function. I'm not sure where it's updated, but I confirmed this by making a SOFORT payment in a debugging session and we exit the function here because the status is **On hold**:

https://github.com/woocommerce/woocommerce-gateway-stripe/blob/c06994c9891151fd1c8975acfead97c55d91993e/includes/class-wc-stripe-webhook-handler.php#L213-L215

> Actually, it seems there are more places where we might need to add the filter:

Yeah, you're right 🤔

I wasn't sure how to test these yesterday and thought these places wouldn't be appropriate, but I figured out how to trigger this now, and the order status hasn't been updated so the custom status isn't taken into account.

Nice catch!

Commit:
Allow custom status when payment processing fails

Made possible via 'wc_stripe_allowed_payment_processing_statuses'
filter.

Before this commit a merchant would not be alerted if a payment fails
when the order has a custom status. By adding the custom status to the
list of allowed statuses through the aforementioned filter the merchant
is alerted when a payment fails.

Note that this is for unusual failures such as the Stripe API being
unavailable. This does not cover "expected" errors such as when a
customer doesn't have enough funds in their account.

reykjalin commented:
@asumaran fixed in 0717af55459cd4813dc79e7453fabcda09af8643 👍

asumaran reviewed (approved):
LGTM!
